### PR TITLE
make thread pool size can be configured for loadqueuepeon

### PR DESCRIPTION
**Pull request by QiuMM**
In my scenario, the KIS tasks would produce thousands of segments at the end of every hour. It will take long time to load the segments, so I want to set the thread pool size for the `loadqueuepeon` to load/drop segments concurrently. 

This PR also modified the `CuratorLoadQueuePeon#processSegmentChangeRequest` since several threads may enter this method and consider `currentlyProcessing == null` at the same moment.

### CODE REVIEW

**Comment by jihoonson** [comment]
@QiuMM would you elaborate more on how this change helps to speed up the segment assignment?

**Comment by QiuMM** [comment]
Well, currently all `loadqueuepeon` share the same `processingExecutor` which contains only one thread so that these `loadqueuepeon` have to process one by one.  Giving the ability to set the thread pool size for the `loadqueuepeon` can let `loadqueuepeon` run in parallel.

**Comment by drcrallen** [comment]
Is the behavior during leadership changes affected by this patch?

**Comment by QiuMM** [comment]
@drcrallen no. This patch just make the thread pool size(default size is one) can be configured, didn't change anything related to leadership.

**Comment by clintropolis** [comment]
Do you have any numbers to compare the differences in segment load throughput of this approach compared to existing single threaded approach? Would make it easier to :+1: this PR I think.

File: services/src/main/java/org/apache/druid/cli/CliCoordinator.java [hunk 1]
```diff
@@ -233,10 +233,14 @@ public LoadQueueTaskMaster getLoadQueueTaskMaster(
               ZkPathsConfig zkPaths
           )
           {
+            int poolSize = Integer.parseInt(properties.getProperty(
```
Comment (egor-ryashin) [inline services/src/main/java/org/apache/druid/cli/CliCoordinator.java]: Could we place that parameter into `DruidCoordinatorConfig`?

File: server/src/main/java/org/apache/druid/server/coordinator/CuratorLoadQueuePeon.java [hunk 2]
```diff
@@ -223,24 +223,26 @@ public void unmarkSegmentToDrop(DataSegment dataSegment)
 
   private void processSegmentChangeRequest()
   {
-    if (currentlyProcessing != null) {
-      log.debug(
-          "Server[%s] skipping processSegmentChangeRequest because something is currently loading[%s].",
-          basePath,
-          currentlyProcessing.getSegmentIdentifier()
-      );
+    synchronized (lock) {
```
Comment (egor-ryashin) [inline server/src/main/java/org/apache/druid/server/coordinator/CuratorLoadQueuePeon.java]: I think it won't work, as all the threads will be checking `currentlyProcessing` until the first thread will reset it (set it to `null`), which will be after the historical node will take and confirm the segment.

File: services/src/main/java/org/apache/druid/cli/CliCoordinator.java (hunk 1 above)
Comment (QiuMM) [reply services/src/main/java/org/apache/druid/cli/CliCoordinator.java]: This parameter would be just used here and wouldn't be called in other classes, so there is no need to hold it in `DruidCoordinatorConfig`.

**Comment by QiuMM** [comment]
@clintropolis It did accelerate the process of load/drop segments when using several threads to run concurrently. But I don't have an exact number to told you how faster it is. Actually, we finally found the real bottleneck is computing the best server to load a segment. Using `CachingCostBalancerStrategy` can help that, but if there are millions of segments it always fallback to `CostBalancerStrategy` because of *serverView is not initialized yet* when I restart the coordinator. Maybe I'll close this PR since this is not the bottleneck and I can not guarantee if there are any other potential issues when configure `processingExecutor` to use multiple threads.

File: server/src/main/java/org/apache/druid/server/coordinator/CuratorLoadQueuePeon.java (hunk 2 above)
Comment (egor-ryashin) [reply server/src/main/java/org/apache/druid/server/coordinator/CuratorLoadQueuePeon.java]: Got it, initially, I thought you'd like to parallelize processing within a single queue peon also.

**Comment by gianm** [comment]
IMO it does make sense to close this if it's not actually the bottleneck, since we don't want to have options lying around that aren't useful (they confuse people and can cause bugs). Does that sound ok @QiuMM?

**Pull request closed without merging**